gax/src/main/java/com/google/api/gax/batching/BatcherStats.java

```diff
@@ -0,0 +1,194 @@
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.SettableApiFuture;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.StatusCode;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.Nullable;
+
+/**
+ * This class keeps the statistics about failed operations(both at RPC and ElementT level) in {@link
+ * Batcher}. This provides the count of individual exception failure and count of each failed {@link
+ * StatusCode.Code} occurred in the batching process.
+ */
+class BatcherStats {
+
+  private final Map<Class, Integer> requestExceptionCounts = new ConcurrentHashMap<>();
+  private final Map<StatusCode.Code, Integer> requestStatusCounts = new ConcurrentHashMap<>();
+  private final AtomicInteger partialBatchFailures = new AtomicInteger(0);
+  private final Map<Class, Integer> entryExceptionCounts = new ConcurrentHashMap<>();
+  private final Map<StatusCode.Code, Integer> entryStatusCounts = new ConcurrentHashMap<>();
+
+  private final Object lock = new Object();
+
+  synchronized void recordBatchFailure(Throwable throwable) {
+    Class exceptionClass = throwable.getClass();
+
+    synchronized (lock) {
+      if (throwable instanceof ApiException) {
+        StatusCode.Code code = ((ApiException) throwable).getStatusCode().getCode();
+        exceptionClass = ApiException.class;
+
+        if (requestStatusCounts.containsKey(code)) {
+          Integer codeCount = requestStatusCounts.get(code);
+          requestStatusCounts.put(code, ++codeCount);
+
+        } else {
+          requestStatusCounts.put(code, 1);
+        }
+      }
+
+      if (requestExceptionCounts.containsKey(exceptionClass)) {
+        Integer exCount = requestExceptionCounts.get(exceptionClass);
+        requestExceptionCounts.put(exceptionClass, ++exCount);
+      } else {
+        requestExceptionCounts.put(exceptionClass, 1);
+      }
+    }
+  }
+
+  <T> void recordBatchElementsCompletion(List<SettableApiFuture<T>> batchElementResultFutures) {
+    final AtomicBoolean markBatchFailure = new AtomicBoolean();
+
+    for (ApiFuture<T> elementResult : batchElementResultFutures) {
+      ApiFutures.addCallback(
+          elementResult,
+          new ApiFutureCallback<T>() {
+            @Override
+            public void onFailure(Throwable throwable) {
+              if (markBatchFailure.compareAndSet(false, true)) {
+                partialBatchFailures.incrementAndGet();
+              }
+              Class exceptionClass = throwable.getClass();
+
+              synchronized (lock) {
+                if (throwable instanceof ApiException) {
+                  StatusCode.Code code = ((ApiException) throwable).getStatusCode().getCode();
+                  exceptionClass = ApiException.class;
```

Comment:
add a comment to explicitly call out that all ApiException subclasses are being aggregated